File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.cpp
class Solution {
public:
    int countUnguarded(int m, int n, vector<vector<int>>& guards,
                       vector<vector<int>>& walls) {
        vector<vector<int>> grid(m, vector<int>(n, 0));

        for (auto& w : walls)
            grid[w[0]][w[1]] = 1;
        for (auto& g : guards)
            grid[g[0]][g[1]] = 2;

        vector<pair<int, int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (auto& g : guards) {
            for (auto& [dx, dy] : dirs) {
                int x = g[0] + dx;
                int y = g[1] + dy;
                // Move until blocked
                while (x >= 0 && x < m && y >= 0 && y < n && grid[x][y] != 1 &&
                       grid[x][y] != 2) {
                    if (grid[x][y] == 0)
                        grid[x][y] = 3; // mark as guarded
                    x += dx;
                    y += dy;
                }
            }
        }

        int count = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (grid[i][j] == 0)
                    count++;

        return count;
    }
};

File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.java
import java.util.*;

class Solution {
    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {
        // Create BitSets for each row to track walls, guards, and guarded cells
        BitSet[] wall = new BitSet[m];
        BitSet[] guard = new BitSet[m];
        BitSet[] guarded = new BitSet[m];
        for (int i = 0; i < m; i++) {
            wall[i] = new BitSet(n);
            guard[i] = new BitSet(n);
            guarded[i] = new BitSet(n);
        }

        // Mark walls and guards
        for (int[] w : walls)
            wall[w[0]].set(w[1]);

        for (int[] g : guards)
            guard[g[0]].set(g[1]);

        // Directions: up, down, left, right
        int[][] dirs = {{-1,0},{1,0},{0,-1},{0,1}};

        // Mark guarded cells
        for (int[] g : guards) {
            int gx = g[0], gy = g[1];
            for (int[] d : dirs) {
                int x = gx + d[0];
                int y = gy + d[1];
                // Keep moving until wall or another guard blocks view
                while (x >= 0 && x < m && y >= 0 && y < n 
                        && !wall[x].get(y) && !guard[x].get(y)) {
                    guarded[x].set(y);
                    x += d[0];
                    y += d[1];
                }
            }
        }

        // Count unguarded cells efficiently
        int unguarded = 0;
        for (int i = 0; i < m; i++) {
            // Combine wall + guard + guarded into one BitSet
            BitSet occupied = (BitSet) wall[i].clone();
            occupied.or(guard[i]);
            occupied.or(guarded[i]);
            // unguarded cells = total - occupied
            unguarded += n - occupied.cardinality();
        }

        return unguarded;
    }
}
